reactor-core/src/main/java/reactor/core/publisher/MonoCollectList.java

```diff
@@ -93,42 +89,92 @@ public void onSubscribe(Subscription s) {
 
 		@Override
 		public void onNext(T t) {
-			if (collection == null) {
+			List<T> c = collection;
+			if (c == TERMINATED) {
 				Operators.onNextDropped(t, actual.currentContext());
 				return;
 			}
-			collection.add(t);
+			//limit the possibilities of triggering concurrent discard + add
+			if (c == CANCELLED) {
+				Operators.onDiscard(t, actual.currentContext());
+				return;
+			}
+
+			c.add(t);
```

Comment:
If you are using the c reference you don't need a Discard List class (no race).

Reply:
can you have a look at the test loops? there were iterations where discarding list would get something added and discard